File: Data Structures & Algorithms/house-robber-ii/submission-15.java
Integer[] dp;
class Solution {
    public int rob(int[] nums) {
        if(nums.length<=2){
            return Arrays.stream(nums).max().getAsInt();
        }
        dp = new Integer[nums.length+1];
        var cost1 = dfs(0,Arrays.copyOfRange(nums,0,nums.length-1));
        dp = new Integer[nums.length+1];
        return Math.max(cost1,dfs(0,Arrays.copyOfRange(nums,1,nums.length)));
        
    }

    public int dfs(int i, int[] nums){
        if(i>=nums.length){
            return 0;
        }
        if(dp[i]!=null){
            return dp[i];
        }
        dp[i] = Math.max(nums[i]+dfs(i+2,nums),dfs(i+1,nums));
        return dp[i];
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-15.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        N = len(nums)
        def dfs(i,j):
            if i == N:
                return 0
            LIS = dfs(i+1,j)
            if j == -1 or nums[i]>nums[j]:
                LIS = max(LIS,1+dfs(i+1,i))
            return LIS
        return dfs(0,-1)
            
        